### PR TITLE
Feature - Finding Empty Slots in Schedule

### LINKED ISSUE #198: Feature - Search for Free Appointment Slots
Add a command to search for empty spaces in the appointment schedule.

### PR DESCRIPTION
**Pull request by sharpstorm**
## Changes
This PR adds the `freebetween` command to search for free appointment slots between a specific time period. It also refactors the Appointments model to allow the display of empty slots in between appointments.

## Known Issues
I will resolve these issues in the meantime, but I expect them to be non-trivial, so when this PR is merged, unsolved bugs will be created as issues.

- ✅ :`freebetween` does not display the correct appointment indexes if a slot appears in between
- ✅ `freebetween` misses out empty slots before the first appointment in the schedule and after the last appointment in the schedule.

## Related Issues
- Closes #198

### CODE REVIEW

**Review by HanJiyao — APPROVED** [review]
Very large changes effort and LGTM.
(Just notice some small name error in Appointment model equals should be otherAppointment not otherPerson as not in changes cannot highlight.)

File: src/main/java/seedu/contax/model/appointment/Appointment.java [hunk 1]
```diff
@@ -6,12 +6,14 @@
 import java.time.LocalDateTime;
 import java.util.Objects;
 
+import seedu.contax.model.chrono.ScheduleItem;
 import seedu.contax.model.person.Person;
 
 /**
- * Represents an appointment in the schedule.
+ * Represents an appointment in the Schedule. Time related functionality is implemented in the superclass
```
Comment (dandaandaaaaaan) [inline src/main/java/seedu/contax/model/appointment/Appointment.java:13]:
```suggestion
 * Represents an appointment in the Schedule. Time related functionality is implemented in the superclass.
```

File: src/main/java/seedu/contax/ui/appointment/AppointmentSlotCard.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+     * Translates a numeric minutes into a more human-readable format.
+     *
+     * @param minutes The number of minutes to translate.
+     * @return A string representing the supplied number of minutes.
+     */
+    private String getReadableDuration(long minutes) {
+        long computedHours = minutes / 60;
+        long computedMinutes = minutes % 60;
+
+        if (computedHours > 2400) {
```
Comment (Wei-Xinn) [inline src/main/java/seedu/contax/ui/appointment/AppointmentSlotCard.java:87]: 10 days is 240 hours bah

**Review by Wei-Xinn — CHANGES_REQUESTED** [review]

**Review by dandaandaaaaaan — CHANGES_REQUESTED** [review]
minor nits, otherwise lgtm. one more thing i spotted while manually testing
Screenshot 2022-03-24 at 7 07 55 PM

the text `Hours` got `s` even though it's 1 hour only, very minor if u dont want fix its okay also

File: src/main/java/seedu/contax/logic/Logic.java [hunk 3]
```diff
@@ -35,8 +35,8 @@
     /** Returns an unmodifiable view of the filtered list of persons */
```
[the remark this answers is not in the document]
Comment (sharpstorm) [reply src/main/java/seedu/contax/logic/Logic.java]: Thanks for spotting it, fixed in 827b102.

File: src/main/java/seedu/contax/logic/Logic.java [hunk 4]
```diff
@@ -35,8 +35,8 @@
...
     ObservableList<Person> getFilteredPersonList();
 
-    /** Returns an unmodifiable view of the list of appointments */
-    ObservableList<Appointment> getAppointmentList();
+    /** Returns an unmodifiable view of the list of appointments and highlighted empty slots */
+    ObservableList<ScheduleItem> getScheduleItemList();
 
     /** Returns an unmodifiable view of the filtered list of tags */
```
[the remark this answers is not in the document]
Comment (sharpstorm) [reply src/main/java/seedu/contax/logic/Logic.java]: Thanks for spotting it, fixed in 827b102.

File: src/main/java/seedu/contax/model/Model.java [hunk 5]
```diff
@@ -185,4 +188,24 @@
      */
     void updateFilteredAppointmentList(Predicate<Appointment> predicate);
 
+    /** Returns an unmodifiable view of the slots between appointments to be highlighted */
```
[the remark this answers is not in the document]
Comment (sharpstorm) [reply src/main/java/seedu/contax/model/Model.java]: Thanks for spotting it, fixed in 827b102.

File: src/main/java/seedu/contax/model/appointment/DisjointAppointmentList.java [hunk 6]
```diff
@@ -242,4 +287,46 @@ private void shiftAppointmentToPosition(int index) {
...
+
+            long minutesBetween = gapStart.until(gapEnd, ChronoUnit.MINUTES);
+            if (minutesBetween >= minimumDuration) {
+                resultList.add(new TimeRange(gapStart, gapEnd));
+            }
+        }
+    }
+
+    private LocalDateTime getEarlier(LocalDateTime dateTime1, LocalDateTime dateTime2) {
+        if (dateTime1.isAfter(dateTime2)) {
```
[the remark this answers is not in the document]
Comment (sharpstorm) [reply src/main/java/seedu/contax/model/appointment/DisjointAppointmentList.java:325]: Thanks for the suggestion, I added it in 827b102.

File: src/main/java/seedu/contax/model/appointment/DisjointAppointmentList.java [hunk 7]
```diff
@@ -242,4 +287,46 @@ private void shiftAppointmentToPosition(int index) {
...
+            return dateTime2;
+        }
+        return dateTime1;
+    }
+
+    private LocalDateTime getLater(LocalDateTime dateTime1, LocalDateTime dateTime2) {
+        if (dateTime1.isAfter(dateTime2)) {
```
[the remark this answers is not in the document]
Comment (sharpstorm) [reply src/main/java/seedu/contax/model/appointment/DisjointAppointmentList.java:337]: Thanks for the suggestion, added in 827b102.

File: src/main/java/seedu/contax/ui/appointment/AppointmentSlotCard.java (hunk 2 above)
Comment (sharpstorm) [reply src/main/java/seedu/contax/ui/appointment/AppointmentSlotCard.java]: I completely missed this thank you for spotting it. Fixed in 827b102.

**Review by Wei-Xinn — APPROVED** [review]
Lgtm

File: src/main/java/seedu/contax/model/appointment/Appointment.java (hunk 1 above)
Comment (sharpstorm) [reply src/main/java/seedu/contax/model/appointment/Appointment.java]: Thanks for spotting it, fixed in 827b102.

File: src/main/java/seedu/contax/logic/commands/FreeBetweenCommand.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            + PREFIX_DATE_START + "STARTDATE "
+            + PREFIX_TIME_START + "STARTTIME "
+            + PREFIX_DATE_END + "ENDDATE "
+            + PREFIX_TIME_END + "ENDTIME "
+            + PREFIX_DURATION + "DURATION* \n"
+            + "Example: `" + COMMAND_WORD + " "
+            + PREFIX_DATE_START + "22-10-2022 "
+            + PREFIX_TIME_START + "12:30 "
+            + PREFIX_DATE_END + "22-10-2022 "
+            + PREFIX_TIME_END + "16:30"
```
[the remark this answers is not in the document]
Comment (sharpstorm) [reply src/main/java/seedu/contax/logic/commands/FreeBetweenCommand.java]: Thanks for spotting it, I also fixed the singular and plural hours/minutes text display in in 827b102.

**Review by dandaandaaaaaan — APPROVED** [review]
lgtm!

**Pull request merged**